### DISCUSSION TITLE
How do you debug pure-python packages in pyodide?

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4895] [Ideas]
## Problem

When trying to develop a python package, it is hard to debug interactively. I tried python-hmr but that experience is still hard to compete with those frontend frameworks'.

For me personally, when I was trying to modify `micropip`, my routine is:

1. modify the code
2. build the wheel (may use `watchfiles` to automate this)
3. start a static file server to serve the `.whl`
4. in the pyodide repl, `pyodide.loadPackage` this wheel in the devtools
5. run test python code on the web page

That's so slow, wasting much time. I know I should write a test case with `pytest-pyodide`, but that doesn't help much.

> (Furthermore, I personally think dev experience is the barrier that makes python ecosystem not that active as JavaScript's)

## Main Idea

In JavaScript, people usually dev with "hmr", which makes you can see the new result as soon as you modify the code.

So I am thinking of making a package like `jupyter`, it behaves like `vite`, which will start a dev server and run your module **instantly** (without the latency to build it if your package is a pure python one like `micropip`) and **reactively** (everytime you save your code, the result updates). And you don't need to reload pyodide because the pyodide instance is cached, and all sources are executed in a layered context instead of `globals()`.

I post this for seeking any thoughts about this or are there already similar tools available.

## Demo

Here is [a stackblitz demo](https://stackblitz.com/edit/pyodide-issue-demo-bizj65?file=src%2FMain.svelte), in which you can see how every keystroke triggers a rerun smoothly:

https://github.com/pyodide/pyodide/assets/74518716/9deac2f6-a6cd-4243-b7c0-2eeea8e44fbe

Note that the cold start may take less than 1 minute usually.

---

About the _running a module without building a wheel_ stuffs, the principle is just like editable installs and an implementation is simple like this: [`load-sources.py`](https://github.com/promplate/pyth-on-line/blob/100dcecfe2da3636cd6414d8e91618383d253a97/src/lib/pyodide/load-sources.py)
Inside that file, the `sources` is a dict, whose keys are paths and values are file contents. The sources can be  easily imported in vite through a simple vite plugin like this: [`dir2json.ts`](https://github.com/promplate/pyth-on-line/blob/100dcecfe2da3636cd6414d8e91618383d253a97/src/lib/plugins/dir2json.ts)

## Further thoughts

- I think it's possible to build a site for people to share interactive python snippets, like codepen for JavaScript.
- I think it's possible to build a site for PyPI, to try any compatible python package directly on the page, like deno/esm.sh playground.

**Comment by ryanking13** [comment]
Yeah, there are some clumsy parts of developing a Python application for Pyodide that especially require browser API access. For now, there aren't very many convenient ways to do it. So, I would appreciate it if anyone could share their experience.

FWIW, Here are some "tricks" or "tips" I use:

1. Separate the parts that require Web API or JavaScript access from the Python code as much as possible so that only the Python code can be tested separately without the browser.
2. Utilize `pytest-pyodide`. It is a useful tool for testing the wheel in the browser without manually running the browser.
  a. @joemarshall made some APIs to copy the local files to browser filesystem, though I haven't tried it myself.
  b. In micropip, we made a pytest fixture to build the local micropip and copy it to the Pyodide distribution, so it can be easily loaded with `pyodide.loadPackage('micropip')`.
3. [Native File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) is often useful for syncing the local file system state with the browser.

---

> which will start a dev server and run your module instantly (without the latency to build it if your package is a pure python one like micropip) and reactively (everytime you save your code, the result updates). And you don't need to reload pyodide because the pyodide instance is cached, and all sources are executed in a layered context instead of globals().

Sounds interesting. What do you think the final form will be, do you think it will work as a standalone tool or as a plugin to other bundlers (vite, etc)?

**Reply by CNSeniorious000** [reply]
Thanks @ryanking13 for sharing these. I think (1) and (2) are best practises and (3) is useful indeed.

And using `watchfiles` and `pytest` together can provide the "rerun-on-save" functionality. There is an old pypi package `pytest-watch` which did this (but have some bug, and I prefer using watchfiles CLI interface instead).

`pytest` + `watchfiles` have at least 2 limitations:

1. Python startup time: everytime `watchfiles` invokes a rerun, pytest is started in a separate process, which is a waste of time
2. No partial rerun. I refer to [`vitest`](https://vitest.dev/guide/features.html)'s ["_only rerun tests that is related to the diff_"](https://x.com/antfu7/status/1468233216939245579)

For `pytest-pyodide`, there is one more:

3. Starting a new selenium session for each function is slow and sometimes may encounter weird errors. And I personally think that a reusing strategy may be hard to implement

I think solution for (1) is easy. If we don't want to start a new python every time a file changes, we can invoke `pytest.main` in a long-lived python process. It can be instantly for most pure-python test cases through this way.

<details><summary>Here is a snippet that can be directly pasted into the pyodide repl:</summary>
<p>

```py
from pathlib import Path
from itertools import count
from asyncio import sleep
import pytest

for i in count():
    Path(f"./test_{i}.py").write_text(f"def test_{i}():\n assert True")
    pytest.main(["--color=no"])
    await sleep(0.01)
```

This prototype is not that fast though, but still much faster than `watchfiles` + `pytest`.

</p>
</details>

For (2), I think that's really hard (at least for me). It requires playing with AST and building a dependency graph. But I think that's possible too.

Of course, maybe these two ideas are uncoupling with `pyodide` and maybe I should post these elsewhere.

For (3), I think strictly isolated environment is useful for CI but is not efficient for developing. My ideal developing experience is like this:

1. Open a pyodide-compatible package workspace folder, let's take `micropip` for instance
2. Run some command, (maybe name it `pyodide dev`?), which behaves like `jupyter lab` — starting a server and have a notebook-like UI
3. Write test code in the UI, and it behaves just like `jupyterlite` notebooks
4. Modify code in my IDE, whenever I save a file, the test code reruns
5. After debugging, I can export that notebook directly into a testcase, (joining all the code blocks into a python function, and I can copy-paste it into a certain `test_*.py`) (and even it can automatically generates assertion statements based on the runned results

Some implementation details:

- I don't know how to _unload_ an imported python package. `importlib.invalidatecache` doesn't work. My thought is to rename the package into a random name, and patch every `import` statement in the test code
- Native File System API is good, but maybe not suitable for this case. Because user maybe not starting the dev server from local machine. I mean, they may start the server from a remote server, a GitHub codespace, or even a static site. And another reason is that Native File System API can't detect file changes.

> What do you think the final form will be

My imagination is like a mixture of jupyterlite + stackblitz + vitest. Specifically, jupyterlite's notebook UI combined with stackblitz's reactivity, resulting the developing experience like vitest.

> Do you think it will work as a standalone tool or as a plugin to other bundlers (vite, etc)

Well, I think that is a standalone tool, but of cource it will integrate with the existing ecosystem. It may be able to export pyodide-pytest test cases, may support pytest fixtures, may be extensible or have programmatic APIs. But I think it will not be a vite server. Because 1. that will require bundling a node executable into the package, which is not elegant. `pyright` is using this way but I don't like it. 2. I think we don't need it. `vite`'s [HMR](https://vitejs.dev/guide/features#hot-module-replacement) interface is suitable for frontend components, it may be possible to make a headless-pyodide-environment a JavaScript module and support HMR for python packages, but I think that's hard and not worthy. I think our module replacement should only uses python and the rerun can use `watchfiles`'s programmatic API. So I think we don't need vite for runtime. The frontend stuff of this project will use vite, but after bundling they are just plain JavaScript files without any JavaScript dependencie.

**Reply by ryanking13** [reply]
Thanks for sharing your detailed ideas. I very much agree with the need for a more comfortable way to test applications or libraries that use Pyodide. However, due to the bandwidth limitations of the core developers, we've been leaving these things to the community. If you were able to come up with some useful solution, we could add your tool to the pyodide organization (of course, if you wish).

**Reply by CNSeniorious000** [reply]
Here are some updates @ryanking13. I've built a stackblitz template [repo](https://github.com/CNSeniorious000/pyodide-minimal-reproduction), and you can have a look:

[![open_in_stackblitz.svg](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/fork/github/CNSeniorious000/pyodide-minimal-reproduction/tree/main-locked?file=python/__main__.py&title=pyodide-reproduction)

This link should bring you to a `main.py`, where the reproduction code should be written.

<details><summary>Note that if you want to have the result updated as soon as you type a char, you have to change the settings here:</summary>
<p>

![image](https://github.com/pyodide/pyodide/assets/74518716/28c8e51f-6335-44bf-9676-7fa7355c17ae)

I've reported this to stackblitz: https://github.com/stackblitz/docs/discussions/223

</p>
</details>

---

<details>
<p>

I also had some CI script to create lockfile from the main branch and to ensure other dependencies to be up-to-date:

[![CI](https://github.com/CNSeniorious000/pyodide-minimal-reproduction/actions/workflows/ci.yml/badge.svg)](https://github.com/CNSeniorious000/pyodide-minimal-reproduction/actions/workflows/ci.yml) [![Auto update](https://github.com/CNSeniorious000/pyodide-minimal-reproduction/actions/workflows/update.yml/badge.svg)](https://github.com/CNSeniorious000/pyodide-minimal-reproduction/actions/workflows/update.yml)

And I think this repository needs no maintenance in the future and I am happy to have it transferred to the @pyodide and we may add this button to the docs.

</p>
</details>

**Reply by ryanking13** [reply]
Cool! Thanks for sharing the demo. Once you think you are ready to share it to the world, how about writing a post about it on the pyodide-blog so that people can try it and give you some feedback?

**Comment by CNSeniorious000** [comment]
Plus, another usage is to generate issue directly from pyodide playground (I called it playground here, which refers to an enhanced pyodide repl-like website)

### The Vision

When users try their code in the repl and encountered an unexpected behavior, they can fire an issue/discussion directly from that site, automatically pasting all the related source code.

And a stackblitz reproduction link is generated too. It would be better if we could build our own stackblitz replacer.

An example is #4901, if you replace `filename="..."` to `filename="<console>"` the result on the right panel will sync in real time.
